### PR TITLE
Added pandas DataFrame and Series parameter types

### PR DESCRIPTION
**Pull request by jlstevens**
Currently WIP. As of the first commit, these two parameter types follow the implementation of `param.Array` which is just syntactic sugar for a `ClassSelector`. I'll now be adding options for additional validation.

### CODE REVIEW

**Comment by jlstevens** [comment]
It looks like pandas isn't playing well with the pypy build. Any suggestions for what I can do about it?

Removing pypy builds is one obvious option, but I would rather not do that.

File: param/__init__.py [hunk 1]
```diff
@@ -1287,6 +1287,73 @@ def __init__(self, **params):
...
+
+        if self.columns is None:
+            pass
+        elif isinstance(self.columns, list):
+            if len(val.columns) != len(self.columns):
+                raise Exception(length_error.format(found=len(val.columns), expected=len(self.columns)))
+            difference = set(val.columns) ^ set(self.columns)
+            if difference:
+                msg = 'Provided DataFrame columns {found} differ from expected columns {expected}'
+                raise Exception(msg.format(found=list(val.columns), expected=self.columns))
```
Comment (jbednar) [inline param/__init__.py]: I would argue for a different approach to validating DataFrame columns.  If someone provides a list of the columns they are expecting from the DataFrame, I think we should only validate that those columns are present.  E.g. if `['x','y']` is specified, it should be treated as a declaration that we will be accessing columns `x` and `y` from the provided dataframe, and so we ensure that those columns exist.  Conversely, if someone specifies cols to be `3`, it's a declaration that we will be accessing columns by number, and that there will need to be three of them, regardless of their names.  

For the numeric case it's not clear to me whether 3 should mean "at least 3", or "exactly 3", as there are arguments both ways. The reason for "at least 3" is that it's expressing specifically what we actually require, i.e. it's a declaration that we will access the first three columns by number; any additional columns won't hurt us but won't be accessed unless we first check that they exist. The reason for "exactly 3" would be to catch errors where someone supplies a larger one but doesn't realize that only the first 3 columns will be used. One could also imagine accepting a tuple to make it feasible to express precisely what you mean (like numeric bounds in Param in general, i.e.. with lower and upper bounds specified explicitly).  Unfortunately, a tuple like that could be ambiguous with the list of columns, e.g. when serialized to JSON (losing the distinction between tuples and lists), so I think I'd favor just picking whether a single numeric argument means "exactly" or "at least".  I think my vote is for "at least".

File: param/__init__.py [hunk 2]
```diff
@@ -1287,6 +1287,73 @@ def __init__(self, **params):
...
+        else:
+            if len(val.columns) != self.columns:
+                raise Exception(length_error.format(found=len(val.columns), expected=self.columns))
+
+        if self.rows is None:
+            pass
+        elif len(val) != self.rows:
+            msg = 'Provided DataFrame has {found} rows which does not match the expected {expected} rows'
+            raise Exception(msg.format(found=len(val), expected=self.rows))
```
Comment (jbednar) [inline param/__init__.py]: For rows, I can easily imagine wanting to express a minimum, maximum, or exact number, which if implemented for rows may as well be implemented for columns with the same syntax if it doesn't conflict with the specific list of columns support.  The current "exactly 3" approach for rows seems unlikely to be sufficient in general.

File: param/__init__.py (hunk 1 above)
Comment (jlstevens) [reply param/__init__.py]: The columns suggestion is now implemented in 3cf180b.

**Comment by jlstevens** [comment]
@ceball I tried editing tox.ini as you suggested in 0c1bab5 but without success. What am I missing?

**Comment by ceball** [comment]
pandas will only be available in matrix entries that include pandas. I.e. the test suite also needs to be able to run without pandas present, by skipping the pandas-requiring tests.

See e.g. https://github.com/ioam/param/blob/master/tests/API1/testnumpy.py#L10

The additional `PARAM_TEST_NUMPY` environment variable is the best hack I could think of at the time to ensure that the numpy tests don't get skipped silently by mistake when they are supposed to be running.

**Comment by ceball** [comment]
You also unfortunately need to add a corresponding entry to travis: https://github.com/ioam/param/blob/master/.travis.yml#L63

This is the kind of pain that inspired pyctdev to start. (So that there could be a different type of pain...)

**Comment by jlstevens** [comment]
@ceball Thanks for that! I think I got things working correctly between bdb4226 and 05caa10...

**Comment by jlstevens** [comment]
Other than adding a few more unit tests, we just need to resolve the semantics we want for numeric rows/columns values. Or naming 'rows' to length. I'm happy to implement things whichever way as long as @jbednar and @philippjfr are in agreement.

File: param/__init__.py (hunk 1 above)
Comment (jbednar) [reply param/__init__.py]: Looks good, thanks.  @philippjfr?

File: param/__init__.py [hunk 3]
```diff
@@ -1287,6 +1287,112 @@ def __init__(self, **params):
         super(Array,self).__init__(ndarray, allow_None=True, **params)
 
 
+class DataFrame(ClassSelector):
+    """
+    Parameter whose value is a pandas DataFrame.
```
Comment (jbednar) [inline param/__init__.py]: Docstring explaining the arguments accepted?

File: param/__init__.py [hunk 4]
```diff
@@ -1287,6 +1287,112 @@ def __init__(self, **params):
...
+                raise Exception(message.format(name=name, length=length, bounds=bounds))
+            else:
+                return
+        (lower, upper) = bounds
+        if lower is not None:
+            if length < lower:
+                raise Exception(message.format(name=name, length=length, bounds=bounds))
+        if upper is not None:
+            if length > upper:
+                raise Exception(message.format(name=name,length=length, bounds=bounds))
```
Comment (jbednar) [inline param/__init__.py]: Probably clearer if expressed a single if statement, with one test expression?

File: param/__init__.py [hunk 5]
```diff
@@ -1287,6 +1287,112 @@ def __init__(self, **params):
...
+
+    def _check_value(self,val,obj=None):
+        super(DataFrame, self)._check_value(val, obj)
+
+        if isinstance(self.columns, set) and self.ordered is True:
+            raise Exception('Columns cannot be ordered when specified as a set')
```
Comment (jbednar) [inline param/__init__.py]: ValueError

File: param/__init__.py (hunk 3 above)
Comment (jlstevens) [reply param/__init__.py]: Addressed in 3dca9a1

File: param/__init__.py (hunk 5 above)
Comment (jlstevens) [reply param/__init__.py]: Addressed in 97c2169

File: param/__init__.py (hunk 4 above)
Comment (jlstevens) [reply param/__init__.py]: Addressed in 70642bc

**Comment by philippjfr** [comment]
Looks good.

**Pull request merged**